llvm/lib/Transforms/Utils/CallPromotionUtils.cpp

```diff
@@ -568,6 +589,46 @@ CallBase &llvm::promoteCallWithIfThenElse(CallBase &CB, Function *Callee,
   return promoteCall(NewInst, Callee);
 }
 
+Constant *llvm::getVTableAddressPointOffset(GlobalVariable *VTable,
+                                            uint32_t AddressPointOffset) {
+  Module &M = *VTable->getParent();
+  const DataLayout &DL = M.getDataLayout();
+  LLVMContext &Context = M.getContext();
+  Type *VTableType = VTable->getValueType();
+  assert(AddressPointOffset < DL.getTypeAllocSize(VTableType) &&
+         "Out-of-bound access");
+  APInt AddressPointOffsetAPInt(32, AddressPointOffset, false);
+  SmallVector<APInt> Indices =
+      DL.getGEPIndicesForOffset(VTableType, AddressPointOffsetAPInt);
```

Comment:
Don't do this. Emit the GEP with i8 element type instead.

Reply:
@nikic thanks for taking a look!

Before I make the change, do you mean `getelementptr i8, ptr %p, i64 40` should be preferred over `getelementptr [[i32 x 10], x 10], ptr %p, i64 0, i64 1, i64 0`  as mentioned in https://groups.google.com/g/llvm-dev/c/U7D6z7ZnKy8?

Reply:
Yes, GEP i8 is canonical if the offset is constant.

Reply:
I found `CallPromotionUtils.cpp` doesn't use `getVTableAddressPointOffset` either, so moved it to `CallPromotionUtilsTest.cpp`